swing-lib/src/test/java/org/triplea/swing/JFrameBuilderTest.java

```diff
@@ -106,27 +110,38 @@ void addComponents() {
   }
 
   @Test
-  void windowClosingEvent() {
-    final AtomicBoolean testValue = new AtomicBoolean(false);
+  void windowClosedAction(@Mock final Runnable action) {
+    final JFrame frame = JFrameBuilder.builder()
+        .windowClosedAction(action)
+        .build();
 
-    JFrameBuilder.builder()
-        .windowCloseAction(() -> testValue.set(true))
-        .build()
-        .dispose();
```

Comment:
window listener events are not always fired, but if we call dispose then we would have always executed the closed action listener. Did you verify that we do *not* rely on that behavior?

Reply:
In other words, do we ever call `dispose` on the battle panel when it is not visible?

Reply:
I'm pretty sure `BattlePanel#battleFrame` is only disposed when the game is shutdown:

https://github.com/triplea-game/triplea/blob/18b417d6584394180dbee974f98f872a2fccf4e2/game-core/src/main/java/games/strategy/triplea/ui/AbstractUiContext.java#L191-L203

I verified `windowClosed()` is called when exiting the game both when `battleFrame` is visible and invisible.

Reply:
Thanks, if that is the case, then does the battle frame close action make sense?

```
PbemDiceRoller.setFocusWindow(null);
```

If the game is shutting down, and we're closing windows, does the focus switch have any effect?

Reply:
I should have stressed the difference between shutting down the _game_ vs. shutting down the _application_.  That is, the difference between the use cases reflected in the **Leave Game** and **Exit Program** commands.

If we only supported the latter use case, then, yes, there would be no need for the battle frame close action.  However, since `PbemDiceRoller#focusWindow` is global state, it probably should continue to be cleared upon closing the battle frame to support the former use case.

Ideally, the solution to get rid of the battle frame close action would be to remove the `PbemDiceRoller#focusWindow` global and come up with cleaner dependency injection and window lifecycle management mechanisms for the battle frame and `PbemDiceRoller`.